File: go/trim.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

// The trim reporter (G3 phase 6, the Go side of ts/src/trim.ts):
// report REDUNDANT map entries — entries whose removal leaves the
// evaluated result unchanged, the spread-implied case included — as
// paths. Report-only, deliberately: canon discards comments and
// layout, so REWRITING the file needs G7's format-preserving patch
// surface.
//
// The test is EVALUATE-AND-COMPARE: for each candidate entry, re-parse
// the source, delete the entry from the parsed tree, evaluate, and
// compare canons. A removal that ERRORS is not redundant: the entry is
// load-bearing. Parsed trees are single-use, so every probe is its own
// parse.

package aontu

import (
	"strings"
)

const (
	TrimClean     = "clean"
	TrimRedundant = "redundant"
	TrimError     = "error"
)

// TrimReport is the whole answer: one verdict, and the redundant paths.
type TrimReport struct {
	Redundant []string `json:"redundant"`
	Verdict   string   `json:"verdict"`
}

// trimCandidates lists every (map, key) pair in a PARSED tree,
// parent-first. List ELEMENTS are not candidates: removing one shifts
// every later index, which is a different document, not the same one
// minus a redundancy. Entries of maps inside lists still are.
func trimCandidates(v Val, path []string, out *[][]string) {
	switch b := v.(type) {
	case *MapVal:
		for _, k := range b.keys {
			kp := append(append([]string{}, path...), k)
			*out = append(*out, kp)
			trimCandidates(b.peg[k], kp, out)
		}
	case *ListVal:
		for i, e := range b.peg {
			trimCandidates(e, append(path, itoa(i)), out)
		}
	}
}

// trimDeleteAt removes the entry at path from a parsed tree. False when
// the path does not address a map entry (which cannot happen for a
// candidate enumerated from an identical parse, but the walk stays
// honest).
func trimDeleteAt(root Val, path []string) bool {
	node := root
	for _, seg := range path[:len(path)-1] {
		switch b := node.(type) {
		case *MapVal:
			node = b.peg[seg]
		case *ListVal:
			i := -1
			for j := range b.peg {
				if itoa(j) == seg {
					i = j
					break
				}
			}
			if i < 0 {
				return false
			}
			node = b.peg[i]
		default:
			return false
		}
		if nil == node {
			return false
		}
	}
	key := path[len(path)-1]
	m, ok := node.(*MapVal)
	if !ok {
		return false
	}
	if _, has := m.peg[key]; !has {
		return false
	}
	delete(m.peg, key)
	keys := m.keys[:0]
	for _, k := range m.keys {
		if k != key {
			keys = append(keys, k)
		}
	}
	m.keys = keys
	opt := m.optional[:0]
	for _, k := range m.optional {
		if k != key {
			opt = append(opt, k)
		}
	}
	m.optional = opt
	return true
}

// trimEvalCanon parses (deleting the entry at delPath, when given),
// evaluates, and answers the canon — with ok=false when the source does
// not stand up, which for the baseline is the caller's error verdict
// and for a probe means "load-bearing".
func (a *Aontu) trimEvalCanon(src string, delPath []string) (string, bool) {
	v, perr := a.parseEntry(src)
	if perr != nil || nil == v {
		return "", false
	}
	if nil != delPath && !trimDeleteAt(v, delPath) {
		return "", false
	}
	ctx := &Ctx{root: v, src: src}
	res := unifyRoot(v, ctx)
	if nil == res || res.Nil() || 0 < len(ctx.err) {
		return "", false
	}
	return res.Canon(), true
}

// TrimCheck is the whole reporter: the baseline canon, then one probe
// per candidate entry, parent-first — and a child of a redundant parent
// is SKIPPED, because "remove the whole entry" already covers it and
// reporting both would tell the author to delete the same text twice.
// Mirrors trimCheck in ts/src/trim.ts.
func (a *Aontu) TrimCheck(src string) TrimReport {
	baseline, ok := a.trimEvalCanon(src, nil)
	if !ok {
		return TrimReport{Verdict: TrimError, Redundant: []string{}}
	}

	parsed, perr := a.parseEntry(src)
	if perr != nil { //coverage:ignore the baseline above already parsed this source
		return TrimReport{Verdict: TrimError, Redundant: []string{}}
	}
	paths := [][]string{}
	trimCandidates(parsed, nil, &paths)

	redundant := []string{}
	for _, path := range paths {
		parent := subPathText(path[:len(path)-1])
		covered := false
		for _, r := range redundant {
			if r == parent || strings.HasPrefix(parent, r+".") {
				covered = true
				break
			}
		}
		if covered {
			continue
		}
		if canon, ok := a.trimEvalCanon(src, path); ok && baseline == canon {
			redundant = append(redundant, subPathText(path))
		}
	}

	verdict := TrimClean
	if 0 < len(redundant) {
		verdict = TrimRedundant
	}
	return TrimReport{Verdict: verdict, Redundant: redundant}
}

File: ts/src/trim.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */
import { includeOpts } from './utility'


import { Aontu } from './aontu'
import { failureFinding } from './vet'
import type { VetFinding } from './vet'
import type { TrustOptions } from './type'


export type TrimVerdict = 'clean' | 'redundant' | 'error'

export type TrimReport = {
  verdict: TrimVerdict
  redundant: string[]

  errors?: VetFinding[]
}

export type TrimOptions = {
  // Where the document CAME FROM, so a relative `@"file"` load inside
  // it resolves from its own directory (vet's schemaPath precedent).
  path?: string
  trust?: TrustOptions

  textExt?: string[]
}


function pathText(path: string[]): string {
  return '$' + (0 < path.length ? '.' + path.join('.') : '')
}


export function candidates(v: any, path: string[], out: string[][]): void {
  if (null == v || true !== v.isVal) {
    return
  }
  if (true === v.isMap && null != v.peg) {
    for (const k of Object.keys(v.peg)) {
      out.push([...path, k])
      candidates(v.peg[k], [...path, k], out)
    }
  }
  else if (true === v.isList && null != v.peg) {
    // List ELEMENTS are not candidates: removing one shifts every
    // later index, which is a different document, not the same one
    // minus a redundancy. Entries of maps INSIDE lists still are.
    for (const k of Object.keys(v.peg)) {
      candidates(v.peg[k], [...path, k], out)
    }
  }
}


// Delete the entry at path from a parsed tree. False when the path
// does not address a map entry (which cannot happen for a candidate
// enumerated from an identical parse, but the walk stays honest).
export function deleteAt(root: any, path: string[]): boolean {
  let node: any = root
  for (const seg of path.slice(0, -1)) {
    if ((true === node?.isMap || true === node?.isList) && null != node.peg) {
      node = node.peg[seg]
    }
    else {
      return false
    }
  }
  const key = path[path.length - 1]
  if (true !== node?.isMap || null == node.peg ||
    !Object.prototype.hasOwnProperty.call(node.peg, key)) {
    return false
  }
  delete node.peg[key]
  // Always an array on a parsed bag (BagVal initialises it).
  node.optionalKeys = node.optionalKeys.filter((k: string) => k !== key)
  return true
}


export function evalCanon(
  src: string, opts: TrimOptions, delPath?: string[],
  sink?: { ctx?: any, failed?: any }): string | undefined {
  const aontu = new Aontu(includeOpts(opts))
  const ctx = aontu.ctx({ collect: true })
  const parseOpts = null == opts.path ? undefined : { path: opts.path }
  const fail = (failed?: any): undefined => {
    if (null != sink) {
      sink.ctx = ctx
      // The failing ROOT travels with the context: a nil root can
      // arrive with an EMPTY error list (`&: id(root)`), and the
      // finding is built from it then (use-cases/BUGS.md §43).
      sink.failed = failed
    }
    return undefined
  }
  const parsed: any = aontu.parse(src, parseOpts, ctx)
  if (0 < ctx.err.length || null == parsed) {
    return fail()
  }
  if (null != delPath && !deleteAt(parsed, delPath)) {
    return undefined
  }
  const v: any = aontu.unify(parsed, parseOpts, ctx)
  if (0 < ctx.err.length || true === v?.isNil) {
    return fail(v)
  }
  return v.canon
}


export function trimCheck(src: string, opts?: TrimOptions): TrimReport {
  const options = opts ?? {}

  const sink: { ctx?: any, failed?: any } = {}
  const baseline = evalCanon(src, options, undefined, sink)
  if (undefined === baseline) {
    return {
      verdict: 'error',
      redundant: [],
      errors: [failureFinding(sink.ctx, options.path, sink.failed)],
    }
  }

  const aontu = new Aontu()
  const ctx = aontu.ctx({ collect: true })
  const parseOpts = null == options.path ? undefined : { path: options.path }
  const parsed: any = aontu.parse(src, parseOpts, ctx)
  const paths: string[][] = []
  candidates(parsed, [], paths)

  const redundant: string[] = []
  for (const path of paths) {
    const parent = pathText(path.slice(0, -1))
    if (redundant.some((r) => r === parent || parent.startsWith(r + '.'))) {
      continue
    }
    if (baseline === evalCanon(src, options, path)) {
      redundant.push(pathText(path))
    }
  }

  return {
    verdict: 0 === redundant.length ? 'clean' : 'redundant',
    redundant,
  }
}
